feat: rewriting cancellation and timeouts page (#4585)
Kotlin/kotlinx.coroutines: docs/topics/cancellation-and-timeouts.md

Cancellation lets you stop a coroutine before it completes.
It stops work that's no longer needed, such as when a user closes a window or navigates away in a user interface while a coroutine is still running.
You can also use it to release resources early and to stop a coroutine from accessing objects past their disposal.

> You can use cancellation to stop long-running coroutines that keep producing values even after other coroutines no longer need them, for example, in [pipelines](channels.md#pipelines).
>
{style="tip"}

Cancellation works through the [`Job`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-job/) handle, which represents the lifecycle of a coroutine and its parent-child relationships.
`Job` allows you to check whether the coroutine is active and allows you to cancel it, along with its children, as defined by [structured concurrency](coroutines-basics.md#coroutine-scope-and-structured-concurrency).

## Cancel coroutines

A coroutine is canceled when the [`cancel()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-job/cancel.html) function is invoked on its `Job` handle.

> This handle is returned by [coroutine builder functions](coroutines-basics.md#coroutine-builder-functions) like

Author: ah 🤔 — I wanted to achieve two things with this sentence: 
- To let the reader know where the `Job` handle comes from 
- To let the reader know that Deferred is (while not in the strictest sense a `Job`) handle it acts the same way in relation to cancellation because it inherits from it

> Also, it's possible to use active voice here, "coroutine builder functions return"

I think so: what do you think about: 

> A coroutine is canceled when the [`cancel()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-job/cancel.html) function is invoked on its `Job` handle.
[Coroutine builder functions](coroutines-basics.md#coroutine-builder-functions) such as
[`.launch()`](coroutines-basics.md#coroutinescope-launch) return a `Job`, and the [`.async()`](coroutines-basics.md#coroutinescope-async)
function returns a [`Deferred`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-deferred/).
`Deferred` is a type that implements `Job` and supports the same cancellation behavior.

the [`.launch()`](coroutines-basics.md#coroutinescope-launch) function, which returns a `Job`, or the [`.async()`](coroutines-basics.md#coroutinescope-async)
function, which returns a [`Deferred`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-deferred/).
`Deferred` is a type that implements `Job` and supports the same cancellation behavior.

You can call the `cancel()` function manually, or it can be invoked automatically through cancellation propagation when a parent coroutine is canceled.

When a coroutine is canceled, it throws a [`CancellationException`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-cancellation-exception/) the next time it checks for cancellation.
For more information about how and when this happens, see [Suspension points and cancellation](#suspension-points-and-cancellation).

> > You can use the [`awaitCancellation()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/await-cancellation.html) function to suspend a coroutine until cancellation.

Reviewer: ```suggestion
> You can use the [`awaitCancellation()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/await-cancellation.html) function to suspend a coroutine until it's canceled.
```

//sampleStart
suspend fun main() {
    withContext(Dispatchers.Default) {
        // Used as a signal that the coroutine has started running
        val job1Started = CompletableDeferred<Unit>()
        
        val job1: Job = launch {
            
            println("The coroutine has started")

            // Completes the CompletableDeferred,
            // signaling that the coroutine has started running
            job1Started.complete(Unit)
            try {
                // Suspends indefinitely
                // Without cancellation, this call would never return
                delay(Duration.INFINITE)
            } catch (e: CancellationException) {
                println("The coroutine was canceled: $e")
              
                // Always rethrow cancellation exceptions!
                throw e
            }
            println("This line will never be executed")
        }
      
        // Waits for job1 to start before canceling it
        job1Started.await()

        // Cancels the coroutine, so delay() throws a CancellationException
        job1.cancel()

        // async returns a Deferred handle, which inherits from Job
        val job2 = async {

>                   // This line may not be printed if the coroutine is canceled before its body starts executing

Reviewer: This comment is tool long, maybe break into two lines like the rest

The coroutine calls `complete()` when it starts executing, and `await()` only returns once that `CompletableDeferred` is completed. This way, cancellation happens only after the coroutine has started running.
The coroutine created by `.async()` doesn't have this check, so it may be canceled before it can run the code inside its block.

> Catching `CancellationException` can break the cancellation propagation.
> If you must catch it, rethrow it to let the cancellation propagate correctly through the coroutine hierarchy.
>
> For more information, see [Coroutine exceptions handling](exception-handling.md#cancellation-and-exceptions).
>
{style="warning"}

### Cancellation propagation

[Structured concurrency](coroutines-basics.md#coroutine-scope-and-structured-concurrency) ensures that canceling a coroutine also cancels all of its children.
This prevents child coroutines from continuing work after the parent has already stopped.

Here's an example:

```kotlin
import kotlinx.coroutines.*
import kotlin.time.Duration

//sampleStart
suspend fun main() {
    withContext(Dispatchers.Default) {

>         // Used as a signal that the child coroutines have been launched

Reviewer: ```suggestion
        // Used as a signal that child coroutines have been launched
```

                println("Child coroutine 1 has started running")
                try {
                    awaitCancellation()
                } finally {
                    println("Child coroutine 1 has been canceled")
                }
            }
            launch {
                println("Child coroutine 2 has started running")
                try {
                    awaitCancellation()
                } finally {
                    println("Child coroutine 2 has been canceled")
                }
            }
            // Completes the CompletableDeferred,
            // signaling that the child coroutines have been launched
            childrenLaunched.complete(Unit)
        }
        // Waits for the parent coroutine to signal that it has launched
        // all of its children
        childrenLaunched.await()

        // Cancels the parent coroutine, which cancels all its children
        parentJob.cancel()
    }
}
//sampleEnd
```
{kotlin-runnable="true" id="cancellation-propagation-example"}

In this example, each child coroutine uses a [`finally` block](exceptions.md#the-finally-block), so the code inside it runs when the coroutine is canceled.

> Here, `CompletableDeferred` signals that the child coroutines are launched before they are canceled, but it doesn't guarantee that they start running. If they are canceled first, nothing is printed.

Reviewer: ```suggestion
Here, `CompletableDeferred` signals that child coroutines are launched before they are canceled, but it doesn't guarantee that they start running. If they are canceled first, nothing is printed.
```

## Make coroutines react to cancellation {id="cancellation-is-cooperative"}

In Kotlin, coroutine cancellation is _cooperative_.
This means that coroutines only react to cancellation when they cooperate by [suspending](#suspension-points-and-cancellation) or [checking for cancellation explicitly](#check-for-cancellation-explicitly).

> In this section, you can learn ways to create cancelable coroutines.

Reviewer: ```suggestion
In this section, you can learn how to create cancelable coroutines.
```

### Suspension points and cancellation

When a coroutine is canceled, it continues running until it reaches a point in the code where it may suspend, also known as a _suspension point_.
If the coroutine suspends there, the suspending function checks whether it has been canceled.

> If it has been canceled, the coroutine stops and throws `CancellationException`.

Reviewer: Repetition, "if it has/if this is the case"

Author: Changed it to : "If it has"

```kotlin
import kotlinx.coroutines.*
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.channels.Channel
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration

suspend fun main() {
    withContext(Dispatchers.Default) {
        val childJobs = listOf(
            launch {
                // Suspends until canceled
                awaitCancellation()
            },
            launch {
                // Suspends until canceled
                delay(Duration.INFINITE)
            },
            launch {
                val channel = Channel<Int>()
                // Suspends while waiting for a value that is never sent
                channel.receive()
            },
            launch {
                val deferred = CompletableDeferred<Int>()
                // Suspends while waiting for a value that is never completed
                deferred.await()
            },
            launch {
                val mutex = Mutex(locked = true)
                // Suspends while waiting for a mutex that remains locked indefinitely
                mutex.lock()
            }
        )
        

>         // Gives the child coroutines time to start and suspend

Author: I think it's better with the:

“the child coroutines” refers to specific coroutines created just above in the code.
Without “the,” to me it sounds more general, as if it applies to any coroutines, not the ones in this example.

>         // Gives the child coroutines time to start and suspend

Reviewer: I think it's clear from the context, the comment is in the middle of the example

#### isActive

Use the [`isActive`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/is-active.html) property in long-running computations to periodically check for cancellation.
This property is `false` when the coroutine is no longer active, which you can use to gracefully stop the coroutine when it no longer needs to continue the operation:

Here's an example:

```kotlin
import kotlinx.coroutines.*
import kotlin.time.Duration.Companion.milliseconds
import kotlin.random.Random

//sampleStart
suspend fun main() {
    withContext(Dispatchers.Default) {
        val unsortedList = MutableList(10) { Random.nextInt() }
        
        // Starts a long-running computation
        val listSortingJob = launch {
            var i = 0

            // Repeatedly sorts the list while the coroutine remains active
            while (isActive) {
                unsortedList.sort()
                ++i
            }
            println(
                "Stopped sorting the list after $i iterations"
            )
        }

>         // Sorts the list for 100 milliseconds, then consider it sorted enough

Reviewer: ```suggestion
        // Sorts the list for 100 milliseconds, then considers it sorted enough
```

suspend fun main() {
    withContext(Dispatchers.Default) {
        val childJob = launch {
            var start = 0
            try {
                while (true) {
                    ++start
                    // Checks the Collatz conjecture for the current number
                    var n = start
                    while (n != 1) {
                        // Throws CancellationException if the coroutine is canceled
                        ensureActive()
                        n = if (n % 2 == 0) n / 2 else 3 * n + 1
                    }
                }
            } finally {
                println("Checked the Collatz conjecture for 0..${start-1}")
            }
        }
        // Runs the computation for one second
        delay(100.milliseconds)

        // Cancels the coroutine
        childJob.cancel()
    }
}
```
{kotlin-runnable="true" id="ensurective-example"}

#### yield()

The [`yield()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/yield.html) function suspends the coroutine and checks for cancellation before resuming.
Without suspending, coroutines on the same thread run sequentially.

> Use `yield` to allow other coroutines to run on the same thread or thread pool before one coroutine finishes:

Reviewer: ```suggestion
Use `yield` to allow other coroutines to run on the same thread or thread pool before one of them finishes:
```

                reader?.close()
            }
        }
    }

    private suspend fun updateUi(reader: BufferedReader) {
        // Shows the file contents
        while (true) {
            val line = withContext(Dispatchers.IO) {
                reader.readLine()
            }
            if (line == null)
                break
            addOneLineToUi(line)
        }
    }

    private fun addOneLineToUi(line: String) {
        // Placeholder for code that adds one line to the UI
    }

    // Only callable from the UI thread
    fun leaveScreen() {
        // Cancels the scope when leaving the screen
        // You can no longer update the UI
        scope.cancel()
    }
}
```

In this example, storing the `BufferedReader` in a variable and closing it in the `finally` block ensures the resource is released even if the coroutine is canceled.

### Run non-cancelable blocks

You can prevent cancellation from affecting certain parts of a coroutine.
To do so, pass [`NonCancellable`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-non-cancellable/) as an argument to the `withContext()` coroutine builder function.

> > Avoid using `NonCancellable` with other coroutine builders like `.launch()` or `.async()`. Doing so breaks structured concurrency by severing the parent-child relationship.

Author: 🤔 I think "severing" more accurately describes what happens here. Disrupt can mean many things; it could mean a temporary disturbance or a slight relationship change. What happens here is that the tree hierarchy is severed/cut between parent and child.

> > Avoid using `NonCancellable` with other coroutine builders like `.launch()` or `.async()`. Doing so breaks structured concurrency by severing the parent-child relationship.

Reviewer: I think this word is out of our doc style, you can use "break" here instead of  "severing", but change the first "breaks" in this sentence to something else, like "breaches, disrupts"
